### PR TITLE
Occ similarity

### PR DESCRIPTION
**Pull request by lizgzil**
---

# Description

New file for occupation aggregates is s3://prinz-green-jobs/outputs/data/ojo_application/extracted_green_measures/analysis/occupation_aggregated_data_20231220_extra.csv

- Find most similar occupations based off the closest cosine distance between a vector made up of proportion of each skill for each occupation
- Notebook to plot these for each occupation
- `aggregate_by_soc_part_2.py` creates an extra column containing the most similar 5 occupations. This is outputted in a `occupation_aggregated_data_{today}_extra.csv` file - the same as before but with one extra column
 
I also output more detailed data on similarities here s3://prinz-green-jobs/outputs/data/ojo_application/extracted_green_measures/analysis/occupation_similarity/20231220/occ_most_similar.json:
- Output data is a dictionary where each key is a SOC 2020 EXT name, and it is a list of the top 10 most similar occupations with some extra information, e.g. for "'Accounting clerks and bookkeepers'"

```
{'Accounting clerks and bookkeepers': [{'SOC_2020_EXT_name': 'Finance administrative occupations n.e.c',
  'similarity': 0.91,
  'number of skills': 67269,
  'average number of skills per job advert': 9.14,
  'popular_overlap': ['issue sales invoices',
   'attend to detail',
   'communication',
   'use spreadsheets software',
   'use microsoft office'],
  'av_proportion_green_skills': 0.0013770632070221,
  'occ_greenness': 'low',
  'ind_greenness': 'high',
  'skills_greenness': 'low',
  'greenness_score': 'low-mid'},
 {'SOC_2020_EXT_name': 'Finance officers',
  'similarity': 0.88,
  'number of skills': 9274,
  'average number of skills per job advert': 9.22,
  'popular_overlap': ['issue sales invoices',
   'attend to detail',
   ...
}
```

I think the important bits are maybe the "SOC_2020_EXT_name" and the "occ_greenness" values.

Note:
I didn't add this to the agg_by_soc script since it was a bit circular, since to calculate this I needed to use the output of agg_by_soc.

Fixes # (issue)

# Instructions for Reviewer

In order to test the code in this PR you need to ...
Screenshot 2023-12-20 at 20 47 46

Please pay special attention to ...

# Checklist:

- [ ] I have refactored my code out from `notebooks/`
- [ ] I have checked the code runs
- [ ] I have tested the code
- [ ] I have run `pre-commit` and addressed any issues not automatically fixed
- [ ] I have merged any new changes from `dev`
- [ ] I have documented the code
  - [ ] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [ ] I have explained this PR above
- [ ] I have requested a code review

### CODE REVIEW

File: dap_prinz_green_jobs/analysis/ojo_analysis/aggregate_by_soc_part_2.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
+"""
+Small script to merge the outputs of aggregate_by_soc.py with occupation_similarity.py
+"""
```
Comment (india-kerle) [inline dap_prinz_green_jobs/analysis/ojo_analysis/aggregate_by_soc_part_2.py]: I would just add this all to aggregate_by_soc.py instead of having it as a separate script?

File: dap_prinz_green_jobs/analysis/ojo_analysis/occupation_similarity.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+    skills_props_per_occ_df = pd.DataFrame(skills_props_per_occ)
+    skills_props_per_occ_df = skills_props_per_occ_df.T
+    skills_props_per_occ_df.fillna(value=0, inplace=True)
+    # Map column names back to their ESCO codes
+    skills_props_per_occ_df.rename(columns=esco_id_2_name, inplace=True)
+    similarities = cosine_similarity(
+        np.array(skills_props_per_occ_df),
+        np.array(skills_props_per_occ_df),
+    )
+    # Get extra information aboutt he occupations using the aggregated by occupation dataset
```
Comment (india-kerle) [inline dap_prinz_green_jobs/analysis/ojo_analysis/occupation_similarity.py]:
```suggestion
    # Get extra information about the occupations using the aggregated by occupation dataset
```

File: dap_prinz_green_jobs/analysis/ojo_analysis/README.md [hunk 3]
```diff
@@ -11,3 +11,22 @@ python dap_prinz_green_jobs/analysis/ojo_analysis/aggregate_by_sic.py #to aggreg
 ```
 
 Meanwhile, the `process_ojo_green_measures.py` file contains methods for analysis. These are largely used in the `notebooks/` directory to generate graphs for the Green Jobs Explorer tool.
+
+### Find similar occupations based of skills asked for
+
+To use the outputs of `/aggregate_by_soc.py` and the skills found across all jobs adverts to find which occupations are most similar to each other, run:
+
+```
+python dap_prinz_green_jobs/analysis/ojo_analysis/occupation_similarity.py
```
Comment (india-kerle) [inline dap_prinz_green_jobs/analysis/ojo_analysis/README.md]: ran perfectly for me!

File: dap_prinz_green_jobs/analysis/ojo_analysis/README.md [hunk 4]
```diff
@@ -11,3 +11,22 @@ python dap_prinz_green_jobs/analysis/ojo_analysis/aggregate_by_sic.py #to aggreg
...
+```
+
+This will create a matrix of the proportions of all skills per occupation, it will then compare these using cosine similarity to find the closest occupations to one another based off which skills are asked for.
+
+We can then join the outputs of this to the original `/aggregate_by_soc.py` output by running:
+
+```
+python dap_prinz_green_jobs/analysis/ojo_analysis/aggregate_by_soc_part_2.py
```
Comment (india-kerle) [inline dap_prinz_green_jobs/analysis/ojo_analysis/README.md]: also no errors - ran perfectly

File: dap_prinz_green_jobs/analysis/ojo_analysis/occupation_similarity.py [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+Get counts and proportions of skills in each occupation.
+Use these to create occupation skill similarity measures.
+"""
+
+import dap_prinz_green_jobs.analysis.ojo_analysis.process_ojo_green_measures as pg
+from dap_prinz_green_jobs.getters.data_getters import save_to_s3, load_s3_data
+from dap_prinz_green_jobs import BUCKET_NAME, PROJECT_DIR, logger, analysis_config
+from dap_prinz_green_jobs.getters.ojo_getters import (
+    get_mixed_ojo_location_sample,
+    get_mixed_ojo_salaries_sample,
```
Comment (india-kerle) [inline dap_prinz_green_jobs/analysis/ojo_analysis/occupation_similarity.py]: i don't think you are using these guys in this script so don't need to import

Comment (india-kerle) [reply dap_prinz_green_jobs/analysis/ojo_analysis/occupation_similarity.py]: VScode has nice visuals to see if you are actually using all your imports so easy to delete

File: dap_prinz_green_jobs/analysis/ojo_analysis/occupation_similarity.py [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+import numpy as np
+from sklearn.metrics.pairwise import cosine_similarity
+from tqdm import tqdm
+
+
+def find_skill_similarity(
+    occ_skills_info: dict,
+    esco_id_2_name: dict,
+    green_esco_id: list,
+    occ_agg_data: str = "outputs/data/ojo_application/extracted_green_measures/analysis/occupation_aggregated_data_20231220_all.csv",
```
Comment (india-kerle) [inline dap_prinz_green_jobs/analysis/ojo_analysis/occupation_similarity.py]: would use the date from the analysis config i.e. `analysis_config["occ_date] (i forget exactly how to call it but you get the gist)` just in case it changes again

File: dap_prinz_green_jobs/analysis/ojo_analysis/occupation_similarity.py [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+
+    today = datetime.now().strftime("%Y%m%d")
+    occ_sim_folder = f"outputs/data/ojo_application/extracted_green_measures/analysis/occupation_similarity/{today}"
+
+    # Save everything needed to calculate occupation similarity based off skills
+    save_to_s3(
+        BUCKET_NAME,
+        esco_id_2_name,
+        f"{occ_sim_folder}/esco_id_2_name.json",
+    )
```
Comment (india-kerle) [inline dap_prinz_green_jobs/analysis/ojo_analysis/occupation_similarity.py]: this is nice - i might use this for new skills if/when we update the graphs!

File: dap_prinz_green_jobs/config/ojo_analysis.yaml [hunk 8]
```diff
@@ -12,3 +12,4 @@ analysis_files:
   agg_soc_date_stamp: "20231220"
   agg_sic_date_stamp: "20231220"
   agg_region_date_stamp: "20231220"
+  occ_most_similar_date: "20231221"
```
Comment (india-kerle) [inline dap_prinz_green_jobs/config/ojo_analysis.yaml:15]: 😎

File: dap_prinz_green_jobs/notebooks/Similar_occupations.ipynb [hunk 9]
```diff
@@ -0,0 +1,337 @@
+{
```
Comment (india-kerle) [inline dap_prinz_green_jobs/notebooks/Similar_occupations.ipynb:1]: great! the most dramatic thing but all the other file names do not have any capital letters so could be saved 'similar_occupations.ipynb`

I took a look at the graphs (love it!) and a few comments:
- i would remove the occupation y axis title as the occupation names often are too long and go over it
- i would use the clean occupation name first to get rid of n.e.c - the function exists in process_ojo_green_measures.py as `clean_soc_name`
- i would rename the occ_greenness legend to "occupation greenness"
- there are a few occupations that only have 1 similar occupation - do we want to filter out those occupations? i don't know what the right answer is because its still interesting so up to you really

**Review by india-kerle — COMMENTED** [review]
thanks liz for the PR! everything ran smoothly for me - i've left a few comments on the data viz 👍

File: dap_prinz_green_jobs/analysis/ojo_analysis/aggregate_by_soc_part_2.py (hunk 1 above)
Comment (lizgzil) [reply dap_prinz_green_jobs/analysis/ojo_analysis/aggregate_by_soc_part_2.py]: Good shout!

File: dap_prinz_green_jobs/analysis/ojo_analysis/occupation_similarity.py (hunk 5 above)
Comment (lizgzil) [reply dap_prinz_green_jobs/analysis/ojo_analysis/occupation_similarity.py]: thanks! I really do need to use VS code!

**Comment by lizgzil** [comment]
thanks so much @india-kerle . I've made those changes - made a lot of sense to put all the soc aggregation in one script. Involved a bit of refactor since it previously relied on a 3 step process, but I'm running it now (I wont merge until it succeeds).
Really weird thing though - I made the changes you suggested to the notebook, added and committed them, but they havent turned up in the PR! no errors though. I'm sure I've done something silly, so will try again with fresh eyes in the morning

**Comment by lizgzil** [comment]
ah ok - the changes to the notebook did work, but changing to the lower case file name didn't!? very weird

**Comment by india-kerle** [comment]
@lizgzil to be fair i think the final name is absolutely reasonable 🥲 - thanks for the quick turn around!
